File: Test/Ml/Data/TestTokenizer.cpp
#include <Oa/Ml/Tokenizer.h>

#include <gtest/gtest.h>

#include <cstdio>
#include <cstring>

namespace {

void ExpectSame(const OaVec<OaI32>& InA, const OaVec<OaI32>& InB) {
	ASSERT_EQ(InA.Size(), InB.Size());
	for (OaUsize i = 0; i < InA.Size(); ++i) EXPECT_EQ(InA[i], InB[i]) << "token " << i;
}

} // namespace

TEST(BpeTokenizer, LearnsFullWidthNestedTokensAndRoundTrips) {
	constexpr const char* corpus =
		"forward strike forward strike backward guard forward strike";
	OaBpeTokenizer tokenizer(320);
	tokenizer.Train(corpus, 64);

	EXPECT_GT(tokenizer.NumMerges(), 0);
	EXPECT_EQ(tokenizer.VocabSize(), 256 + tokenizer.NumMerges());
	const auto encoded = tokenizer.Encode(corpus);
	EXPECT_LT(encoded.Size(), std::strlen(corpus));
	bool hasLearnedToken = false;
	for (OaI32 token : encoded) hasLearnedToken |= token >= 256;
	EXPECT_TRUE(hasLearnedToken);
	EXPECT_EQ(tokenizer.Decode(encoded), corpus);
}

TEST(BpeTokenizer, TrainingIsDeterministic) {
	constexpr const char* corpus = "left right left right fast slow fast slow";
	OaBpeTokenizer a(300), b(300);
	a.Train(corpus, 44);
	b.Train(corpus, 44);
	EXPECT_EQ(a.NumMerges(), b.NumMerges());
	ExpectSame(a.Encode(corpus), b.Encode(corpus));
}

TEST(BpeTokenizer, PersistsExactVocabulary) {
	constexpr const char* corpus = "a fighter steps forward and swings the sword";
	const OaString path = "/tmp/oa_bpe_tokenizer_test.txt";
	OaBpeTokenizer trained(320);
	trained.Train(corpus, 64);
	ASSERT_TRUE(trained.Save(path).IsOk());

	OaBpeTokenizer loaded;
	ASSERT_TRUE(loaded.Load(path).IsOk());
	EXPECT_EQ(loaded.VocabSize(), trained.VocabSize());
	ExpectSame(loaded.Encode(corpus), trained.Encode(corpus));
	EXPECT_EQ(loaded.Decode(loaded.Encode(corpus)), corpus);
	std::remove(path.CStr());
}

File: test/rs/ml/test_tokenizer.rs
use std::{
	fs,
	time::{SystemTime, UNIX_EPOCH},
};

use oa::ml::BpeTokenizer;

#[test]
fn bpe_learns_full_width_nested_tokens_and_round_trips() {
	let corpus = b"forward strike forward strike backward guard forward strike";
	let mut tokenizer = BpeTokenizer::new(320);
	tokenizer.train(corpus, 64);

	assert!(tokenizer.num_merges() > 0);
	assert_eq!(tokenizer.vocab_size(), 256 + tokenizer.num_merges());
	let encoded = tokenizer.encode(corpus);
	assert!(encoded.len() < corpus.len());
	assert!(encoded.iter().any(|token| *token >= 256));
	assert_eq!(
		tokenizer.decode(&encoded).expect("decode should succeed"),
		corpus
	);
}

#[test]
fn bpe_training_and_prompt_alignment_are_deterministic() {
	let corpus = b"left right left right fast slow fast slow";
	let mut left = BpeTokenizer::new(300);
	let mut right = BpeTokenizer::new(300);
	left.train(corpus, 44);
	right.train(corpus, 44);
	assert_eq!(left.merges(), right.merges());
	assert_eq!(left.encode(corpus), right.encode(corpus));

	let prompt = left.encode_prompt(b"left right", 8);
	let encoded = left.encode(b"left right");
	let copied = encoded.len().min(8);
	assert_eq!(&prompt[8 - copied..], &encoded[encoded.len() - copied..]);
	assert!(prompt[..8 - copied].iter().all(|token| *token == 0));
}

#[test]
fn bpe_persists_exact_vocabulary_and_rejects_invalid_dependencies() -> oa::Result<()> {
	let corpus = "a fighter steps forward and swings the sword";
	let unique = SystemTime::now()
		.duration_since(UNIX_EPOCH)
		.expect("system time must follow Unix epoch")
		.as_nanos();
	let path = std::env::temp_dir().join(format!("oa_bpe_{}_{}.txt", std::process::id(), unique));
	let mut trained = BpeTokenizer::new(320);
	trained.train_text(corpus, 64);
	trained.save(&path)?;

	let mut loaded = BpeTokenizer::default();
	loaded.load(&path)?;
	assert_eq!(loaded.merges(), trained.merges());
	assert_eq!(loaded.encode_text(corpus), trained.encode_text(corpus));
	assert_eq!(loaded.decode_text(&loaded.encode_text(corpus))?, corpus);
	fs::remove_file(&path).expect("test vocabulary should remain removable");

	fs::write(&path, "oa_bpe_v1\n1\n256 0\n").expect("invalid fixture should be writable");
	let before = loaded.merges().to_vec();
	let error = loaded
		.load(&path)
		.expect_err("forward merge dependency was accepted");
	assert_eq!(error.kind(), oa::ErrorKind::DataLoss);
	assert_eq!(loaded.merges(), before);
	fs::remove_file(path).expect("invalid fixture should remain removable");
	Ok(())
}
